crates/node/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+                tokio::spawn(cluster_controller.run(component_shutdown_watch.clone()))
+            })
+            .into();
+
+        tokio::select! {
+            _ = &mut shutdown_signal => {
+                drop(component_shutdown_watch);
+                let _ = tokio::join!(component_shutdown_signal.drain(), &mut cluster_controller_handle);
+                return Ok(());
+            },
+            Some(cluster_controller_result) = &mut cluster_controller_handle => {
+                cluster_controller_result.map_err(|err| Error::panic("cluster controller role", err))??;
+                panic!("Unexpected termination of cluster controller role.");
+            },
+            attachment_result = Self::attach_node(self.node_id, self.cluster_controller_endpoint) => {
+                attachment_result?
+            }
+        }
+
+        let mut worker_handle = tokio::spawn(self.worker_role.run(component_shutdown_watch));
+
+        tokio::select! {
+            // todo: node should also run the node-ctrl endpoint and forward signal to components
+            _ = shutdown_signal => {
+                info!("Shutting node down");
+                let _ = tokio::join!(component_shutdown_signal.drain(), worker_handle, cluster_controller_handle);
+            },
+            worker_result = &mut worker_handle => {
+                worker_result.map_err(|err| Error::panic("worker role", err))??;
+                panic!("Unexpected termination of worker role.");
+            },
+            Some(cluster_controller_result) = &mut cluster_controller_handle => {
+                cluster_controller_result.map_err(|err| Error::panic("cluster controller role", err))??;
+                panic!("Unexpected termination of cluster controller role.");
+            },
+        }
+
+        Ok(())
+    }
+
+    async fn attach_node(node_id: NodeId, cluster_controller_endpoint: Uri) -> Result<(), Error> {
+        info!("Attach to cluster at '{cluster_controller_endpoint}'");
+        let channel = Channel::builder(cluster_controller_endpoint.clone())
+            .connect_timeout(Duration::from_secs(5))
+            .connect_lazy();
+        let cc_client = ClusterControllerClient::new(channel);
+
+        RetryPolicy::exponential(Duration::from_millis(50), 2.0, 10, None)
+            .retry_operation(|| async {
+                cc_client
+                    .clone()
+                    .attach_node(AttachmentRequest {
+                        node_id: Some(node_id.into()),
+                    })
+                    .await
+            })
+            .await
+            .map_err(|err| Error::Attachment(cluster_controller_endpoint, err))?;
+
+        Ok(())
+    }
+}
+
+/// Specifying where the cluster controller runs. Options are:
+///
+/// * Local: Spawning the controller as part of this process
+/// * Remote: The controller runs on a remote host
+#[derive(Debug)]
+pub enum ClusterControllerLocation {
+    Local,
+    Remote(String),
+}
```

Comment:
Heads up, this potentially can be swapped by a future type that I'll introduce that looks like this:
```rs
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum NetworkAddress {
    /// Unix domain socket
    Uds(String),
    /// TCP socket address
    Tcp(SocketAddr),
}

impl FromStr for NetworkAddress {
    type Err = NodesConfigError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.starts_with("unix:") {
            Ok(NetworkAddress::Uds(s[5..].to_string()))
        } else {
            Ok(NetworkAddress::Tcp(s.parse()?))
        }
    }
}
```

This will be a direct replacement if you agreed that we can avoid special casing the "local" variant.

Reply:
I think it would be cool to also support unix domain sockets :-) In that case, I think I need to model the roles explicitly because right now the `ClusterControllerLocation` implicitly encodes whether the process is supposed to run the cluster controller or not. Will look into this once we have this `NetworkAddress` type.